rust/nasl-builtin-regex/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+// SPDX-FileCopyrightText: 2024 Greenbone AG
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+use nasl_builtin_utils::{error::FunctionErrorKind, NaslFunction};
+use nasl_builtin_utils::{Context, Register};
+use nasl_function_proc_macro::nasl_function;
+use nasl_syntax::NaslValue;
+use regex::{Regex, RegexBuilder};
+
+fn parse_search_string<'a>(s: &'a mut &str, rnul: bool, multiline: bool) -> &'a str {
+    if !rnul {
+        *s = s.split('\0').next().unwrap();
+    }
+    if !multiline {
+        *s = s.split('\n').next().unwrap();
+    }
+
+    s
+}
```

Comment:
This can be simplified by removing the double reference. This may be a bit confusing but what we want to do is really just change to which part of the original `String` the `&str` should point to, so we don't actually need a mutable reference here - instead we just return a new `&str` that points to the right place. This is not a costly operation since `&str` is a tiny type.

```rust
fn parse_search_string(mut s: &str, rnul: bool, multiline: bool) -> &str {
    if !rnul {
        s = s.split('\0').next().unwrap();
    }
    if !multiline {
        s = s.split('\n').next().unwrap();
    }

    s
}
```

Looks like a small difference, but it means that we don't need to pass `&mut &str` to this function, which simplifies code further downstream. Also since its just a single reference, the compiler can now tell what the lifetime of the returned reference is.